src/native/string_utils.c

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include <jni.h>
+
+#include <aws/common/encoding.h>
+#include <aws/common/string.h>
+
+#include "crt.h"
+
+JNIEXPORT
+jbyteArray JNICALL Java_software_amazon_awssdk_crt_utils_StringUtils_stringUtilsBase64Encode(
+    JNIEnv *env,
+    jclass jni_class,
+    jbyteArray jni_data) {
+    (void)jni_class;
+
+    struct aws_byte_cursor data_cursor;
+    AWS_ZERO_STRUCT(data_cursor);
+    if (jni_data != NULL) {
+        data_cursor = aws_jni_byte_cursor_from_jbyteArray_acquire(env, jni_data);
+    } else {
+        return NULL;
+    }
+
+    // Determine how much space we need
+    size_t terminated_length = 0;
+    if (aws_base64_compute_encoded_len(data_cursor.len, &terminated_length) != AWS_OP_SUCCESS) {
+        goto clean_up;
+    }
+
+    struct aws_byte_buf formatted_data;
+    AWS_ZERO_STRUCT(formatted_data);
+    aws_byte_buf_init(&formatted_data, aws_jni_get_allocator(), terminated_length);
+    int result = aws_base64_encode(&data_cursor, &formatted_data);
+    if (result != AWS_OP_SUCCESS) {
+        aws_byte_buf_clean_up(&formatted_data);
+        goto clean_up;
+    }
+
+    aws_jni_byte_cursor_from_jbyteArray_release(env, jni_data, data_cursor);
+
+    struct aws_byte_cursor formatted_data_cursor = aws_byte_cursor_from_buf(&formatted_data);
+    jbyteArray return_data = aws_jni_byte_array_from_cursor(env, &formatted_data_cursor);
+    aws_byte_buf_clean_up_secure(&formatted_data);
+
+    return return_data;
+
+clean_up:
+    aws_jni_byte_cursor_from_jbyteArray_release(env, jni_data, data_cursor);
+    return NULL;
```

Comment:
I am not a fan of functions with multiple error-handling paths, where each path has slightly different cleanup needs. It's just so hard to get right, and error-handling paths can be very hard to test.

That's why, in our C APIs, we always try to have the clean-up function tolerate being called on something that's zeroed out or null. So, you can just naively try to clean up even if it was never initialized.

I suggest structuring the function like this:
```
jbyteArray JNICALL Java_software_thisFunction(...) {

    /* declare everything that may need cleanup, and zero it out */
    struct aws_byte_buf formatted_data;
    AWS_ZERO_STRUCT(formatted_data)
    jbyteArray returnData = NULL;

    struct aws_byte_cursor data_cursor =  aws_jni_byte_cursor_from_jbyteArray_acquire(env, jni_data);
    if (data_cursor.ptr == NULL) {
        goto cleanup;
    }

    ... blah blah blah if anything goes wrong goto cleanup

    return_data = aws_jni_byte_array_from_cursor(env, &formatted_data_cursor);

    /* hey check it out even the success case falls through to cleanup */

clean_up:
    aws_jni_byte_cursor_from_jbyteArray_release(env, jni_data, data_cursor);
    aws_byte_buf_clean_up(&formatted_data);
    return returnData;
}
```